compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -926,13 +927,51 @@ private boolean isXMLTypeAssignable(BType sourceType, BType targetType, Set<Type
                     return isAssignable(sourceType, target.constraint, unresolvedTypes);
                 }
                 BXMLType source = (BXMLType) sourceType;
+                if (source.constraint.tag == TypeTags.NEVER) {
+                    if (sourceTag == targetTag) {
+                        return true;
+                    }
+                    return isAssignable(source, target.constraint, unresolvedTypes);
+                }
                 return isAssignable(source.constraint, target.constraint, unresolvedTypes);
             }
             return true;
         }
+        if (targetTag == TypeTags.XML_TEXT && sourceTag == TypeTags.XML) {
+            BXMLType source = (BXMLType) sourceType;
+            if (source.constraint != null) {
+                return source.constraint.tag == TypeTags.NEVER;
+            }
+        }
+        if (targetTag == TypeTags.STRING && sourceTag == TypeTags.XML) {
+            BXMLType source = (BXMLType) sourceType;
+            if (source.constraint.tag == TypeTags.NEVER) {
+                return true;
+            }
+            return isAssignable(source.constraint, targetType, unresolvedTypes);
+        }
         return sourceTag == targetTag;
     }
 
+    public boolean isConversionExprXMLNeverUnion(BLangTypeConversionExpr conversionExpr) {
+        if (conversionExpr.expr.type.tag == TypeTags.XML) {
+            BXMLType conversionExpressionType = (BXMLType) conversionExpr.expr.type;
+            //Revisit and check xml<xml<constraint>>> on chained iteration
+            while (conversionExpressionType.constraint.tag == TypeTags.XML) {
+                conversionExpressionType = (BXMLType) conversionExpressionType.constraint;
+            }
+            if (conversionExpressionType.constraint.tag == TypeTags.NEVER ||
+                    conversionExpressionType.constraint.tag == TypeTags.XML_TEXT) {
+                return true;
+            }
+            return false;
+        }
+        if (conversionExpr.expr.type.tag == TypeTags.UNION) {
+            return isAllXMLMembers((BUnionType) conversionExpr.expr.type);
```

Comment:
Shouldn't we be calling `isConversionExprXMLNeverUnion` recursively for all members here?

Reply:
No right we should just check if each member is assignable to xmlTextType ?